File: cpp/FSalgorithms.cpp
#include "FSalgorithms.h"
#include <queue>
#include <vector>
#include <utility> // для std::pair
#include <stack>
using namespace std;

vector<pair<unsigned int, unsigned int>> Neighbours(pair<unsigned int, unsigned int> current, vector<vector<unsigned short>> &maze) {
  vector<pair<unsigned int, unsigned int>> neighbours;
  unsigned int rows = maze.size(); 
  unsigned int cols = maze[0].size();
  auto i = current.first, j = current.second;
  
  if (i > 0 && maze[i-1][j] == 0) {neighbours.push_back({i-1, j});}
  if (i < rows - 1 && maze[i+1][j] == 0) {neighbours.push_back({i+1, j});}
  if (j > 0 && maze[i][j-1] == 0) {neighbours.push_back({i, j-1});}
  if (j < cols - 1 && maze[i][j+1] == 0) {neighbours.push_back({i, j+1});}

  return neighbours;
}

int BFS(pair<unsigned int, unsigned int> start, pair<unsigned int, unsigned int> end, vector<vector<unsigned short>> &maze) {
  if (maze.empty() || maze[0].empty()) {return 0;}
  queue<pair<unsigned int, unsigned int>> queue;
  vector<vector<pair<unsigned int, unsigned int>>> parent;
  vector<vector<bool>> visited(maze.size(), vector<bool>(maze[0].size(), false)); // с помощью конструктора передаем в visited значения: для maze.size() столбцов в каждую клетку передаем вектор булевых значений размером maze[0].size() колонок со значениями false;
  queue.push(start);
  visited[start.first][start.second] = true;

  parent.resize(maze.size());
  for (auto &row : parent) {row.resize(maze[0].size());}
  parent[start.first][start.second] = start;

  while (!queue.empty()) {
    auto current = queue.front();
    queue.pop();

    if (current == end) {
      pair<unsigned int, unsigned int> at = end;
      while (at != start) {
        maze[at.first][at.second] = 2;
        at = parent[at.first][at.second];
      }
      maze[start.first][start.second] = 2;
      return 1;
    }

    for (auto neighbour : Neighbours(current, maze)) {
      if (!visited[neighbour.first][neighbour.second]) {
        visited[neighbour.first][neighbour.second] = true;
        queue.push(neighbour);
        parent[neighbour.first][neighbour.second] = current;
      }
    }
  }
  
  return 0;
}

int DFS(pair<unsigned int, unsigned int> start, pair<unsigned int, unsigned int> end, vector<std::vector<unsigned short>> &maze) {
  if (maze.empty() || maze[0].empty()) {return 0;}
  stack<pair<unsigned int, unsigned int>> dots;
  vector<vector<bool>> visited(maze.size(), vector<bool>(maze[0].size(), false));
  vector<vector<pair<unsigned int, unsigned int>>> parent;
  
  dots.push(start);
  visited[start.first][start.second] = true;

  parent.resize(maze.size());
  for (auto &row : parent) {row.resize(maze[0].size());}
  parent[start.first][start.second] = start;

  while (!dots.empty()) {
    auto current = dots.top();
    dots.pop();

    if (current == end) {
      auto at = end;
      while (at != start) {
        maze[at.first][at.second] = 2;
        at = parent[at.first][at.second];
      }
      maze[start.first][start.second] = 2;
      return 1;
    }
    
    for (auto neighbour : Neighbours(current, maze)) {
      if (!visited[neighbour.first][neighbour.second]) {
        dots.push(neighbour);
        visited[neighbour.first][neighbour.second] = true;
        parent[neighbour.first][neighbour.second] = current;
      }
    }
  }

  return 0;
}

File: go/utils/FSalgorithms.go
package FSalgorithms

func Neighbours(current [2]uint, maze [][]uint) [][2]uint {
    var neighbours [][2]uint
    rows := len(maze)
    if rows == 0 {
        return neighbours
    }
    cols := len(maze[0])
    i, j := current[0], current[1]

    if i > 0 && maze[i-1][j] == 0 {
        neighbours = append(neighbours, [2]uint{i - 1, j})
    }
    if i < rows-1 && maze[i+1][j] == 0 {
        neighbours = append(neighbours, [2]uint{i + 1, j})
    }
    if j > 0 && maze[i][j-1] == 0 {
        neighbours = append(neighbours, [2]uint{i, j - 1})
    }
    if j < cols-1 && maze[i][j+1] == 0 {
        neighbours = append(neighbours, [2]uint{i, j + 1})
    }
    return neighbours
}

func BFS(start, end [2]uint, maze *[][]uint) int {
    if len(*maze) == 0 || len((*maze)[0]) == 0 {
        return 0
    }

    var queue [][2]uint
    parent := make([][][2]uint, len(*maze))
    for i := range parent {
        parent[i] = make([][2]uint, len((*maze)[0]))
    }
    visited := make([][]bool, len(*maze))
    for i := range visited {
        visited[i] = make([]bool, len((*maze)[0]))
    }

    queue = append(queue, start)
    visited[start[0]][start[1]] = true
    parent[start[0]][start[1]] = start

    for len(queue) > 0 {
        current := queue[0]
        queue = queue[1:]

        if current == end {
            at := end
            for at != start {
                (*maze)[at[0]][at[1]] = 2
                at = parent[at[0]][at[1]]
            }
            (*maze)[at[0]][at[1]] = 2
            return 1
        }

        for _, neighbour := range Neighbours(current, *maze) {
            if !visited[neighbour[0]][neighbour[1]] {
                visited[neighbour[0]][neighbour[1]] = true
                queue = append(queue, neighbour)
                parent[neighbour[0]][neighbour[1]] = current
            }
        }
    }
    return 0
}

func DFS(start, end [2]uint, maze *[][]uint) int {
    if len(*maze) == 0 || len((*maze)[0]) == 0 {
        return 0
    }

    var stack [][2]uint
    visited := make([][]bool, len(*maze))
    for i := range visited {
        visited[i] = make([]bool, len((*maze)[0]))
    }
    parent := make([][][2]uint, len(*maze))
    for i := range parent {
        parent[i] = make([][2]uint, len((*maze)[0]))
    }

    stack = append(stack, start)
    visited[start[0]][start[1]] = true
    parent[start[0]][start[1]] = start

    for len(stack) > 0 {
        current := stack[len(stack)-1]
        stack = stack[:len(stack)-1] // Исправлено: было stack = stack[len(stack)-1]

        if current == end {
            at := end
            for at != start {
                (*maze)[at[0]][at[1]] = 2
                at = parent[at[0]][at[1]]
            }
            (*maze)[at[0]][at[1]] = 2
            return 1
        }

        for _, neighbour := range Neighbours(current, *maze) {
            if !visited[neighbour[0]][neighbour[1]] {
                stack = append(stack, neighbour)
                visited[neighbour[0]][neighbour[1]] = true
                parent[neighbour[0]][neighbour[1]] = current
            }
        }
    }
    return 0
}
